source/ranges.tex

```diff
@@ -10040,3 +10070,1603 @@
+\exposid{n_} with \tcode{n}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr auto begin()
+  requires (!(@\exposconcept{simple-view}@<V> && @\exposconcept{slide-caches-nothing}@<const V>));
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\begin{itemize}
+\item
+If \tcode{V} models \exposconcept{slide-caches-first},
+\begin{codeblock}
+@\exposid{iterator}@<false>(ranges::begin(@\exposid{base_}@),
+             ranges::next(ranges::begin(@\exposid{base_}@), @\exposid{n_}@ - 1, ranges::end(@\exposid{base_}@)), @\exposid{n_}@)
+\end{codeblock}
+\item
+Otherwise, \exposid{iterator}<false>(ranges::begin(\exposid{base_}), \exposid{n_}).
+\end{itemize}
+
+\remarks
+In order to provide the amortized constant-time complexity
+required by the \libconcept{range} concept,
+this function caches the result within the \tcode{slide_view}
+for use on subsequent calls
+when \tcode{V} models \exposconcept{slide-caches-first}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr auto begin() const requires @\exposconcept{slide-caches-nothing}@<const V>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{\exposid{iterator}<true>(ranges::begin(\exposid{base_}), \exposid{n_})}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr auto end()
+  requires (!(@\exposconcept{simple-view}@<V> && @\exposconcept{slide-caches-nothing}@<const V>));
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\begin{itemize}
+\item
+If \tcode{V} models \exposconcept{slide-caches-nothing},
+\begin{codeblock}
+@\exposid{iterator}@<false>(ranges::begin(@\exposid{base_}@) + range_difference_t<V>(size()), @\exposid{n_}@)
+\end{codeblock}
+\item
+Otherwise, if \tcode{V} models \exposconcept{slide-caches-last},
+\begin{codeblock}
+@\exposid{iterator}@<false>(ranges::prev(ranges::end(@\exposid{base_}@), @\exposid{n_}@ - 1, ranges::begin(@\exposid{base_}@)), @\exposid{n_}@)
+\end{codeblock}
+\item
+Otherwise, if \tcode{V} models \libconcept{common_range},
+\begin{codeblock}
+@\exposid{iterator}@<false>(ranges::end(@\exposid{base_}@), ranges::end(@\exposid{base_}@), @\exposid{n_}@)
+\end{codeblock}
+\item
+Otherwise, \tcode{\exposid{sentinel}(ranges::end(\exposid{base_})).}
+\end{itemize}
+
+\pnum
+\remarks
+In order to provide the amortized constant-time complexity
+required by the \libconcept{range} concept,
+this function caches the result within the \tcode{slide_view}
+for use on subsequent calls
+when \tcode{V} models \exposconcept{slide-caches-last}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr auto end() const requires slide-caches-nothing<const V>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\returns
+\tcode{begin() + range_difference_t<const V>(size())}.
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr auto size() requires sized_range<V>;
+constexpr auto size() const requires sized_range<const V>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+auto sz = ranges::distance(@\exposid{base_}@) - @\exposid{n_}@ + 1;
+if (sz < 0) sz = 0;
```

Comment:
Looks like `max<`_`T`_`>(0, ...)` for _`T`_ = `decltype(sz)` (current `sz` above).
```suggestion
auto sz = max({}, ranges::distance(@\exposid{base_}@) - @\exposid{n_}@ + 1);
```

Reply:
Not doing this for now; please file a separate editorial issue.

Reply:
That's https://github.com/cplusplus/draft/issues/5250.